tokio/src/runtime/thread_pool/worker.rs

```diff
@@ -543,6 +544,9 @@ impl Core {
     /// Signals all tasks to shut down, and waits for them to complete. Must run
     /// before we enter the single-threaded phase of shutdown processing.
     fn pre_shutdown(&mut self, worker: &Worker) {
+        // Close the OwnedTasks to prevent spawning new tasks during shutdown.
+        worker.shared.owned.close();
+
```

Comment:
Should we do this when closing the inject queue?

Reply:
Probably? I would guess we should close at the same point.

Also, having two "closed" states to update makes me uncomfortable. It seems like a potential source of bugs. Is there a way to have a single closed flag (either the owned set or the inject queue)?

What would happen if we only had `shared.owned` track closed instead of the inject queue? Then, when closing, there are no more tasks that can be spanwed, all existing tasks are "shutdown", then the queues drained? The inject queue should no longer be used I think.

Reply:
Please see the comment in [`deefadc`](https://github.com/tokio-rs/tokio/pull/3955/commits/deefadcec989225e3f584f30d192b78827df857d) for why we need both close bits.